### ISSUE TITLE
Use an external database instead of `.spor`

### ISSUE THREAD

**Issue by aslakhellesoy** [abingham/rust_spor#8]
Hi @abingham

Spor looks very interesting - thanks for writing this library!

I've got a use case where other programs need to query the spor metadata. These programs will have access to repository files via a custom protocol, and they do not have access to the git repository. We could transport the spor metadata as files too, but it would be preferrable to be able to store it in a database such as postgres. A central database could provide quicker and easier access to the metadata, and it also allows users not to "pollute" their repository with metadata.

Would this kind of extension fit in with the design goals of spor? If so, would you be interested in a PR for inclusion in a future release? It would be my first foray into Rust, but from what I can tell your code looks very clean and well tested, so I hope it wouldn't be too hard.

Many thanks,
Aslak Hellesøy

**Comment by abingham** [comment]
I certainly don't object to this idea on any principle. The current design
is largely just the first thing that occurred to me, and I suspect it'll
need to change.

That said, a few issues come to mind immediately. First, the spor
command-line tools currently rely on finding the `.spor` directory to do
their work, e.g. find the metadata, construct relative paths, etc. We'd
need to replace this with something else, e.g. possibly a spor config file.
I think this is largely a technical issue, and not too hard to address.

The second issue is more existential. Who will be in charge of updating the
metadata? If it exists in a central database, and if there are two
developers hacking on the same file, which has the "right" to update the
metadata (thereby putting the other person out of date)? We solve this
right now by giving everyone their own copy of the metadata via git. This
will require some thought.

I'm definitely sympathetic to the idea of extentions/plugins. If spor
offered multiple "storage backend" options, my instinct would be to provide
them as plugins. As I have exactly zero experience with plugins in Rust,
this could be an adventure!

Let me know what you think, and I'll definitely give this some more thought.

**Comment by aslakhellesoy** [comment]
> Who will be in charge of updating the metadata?

In our case, that would be done by a server side process after each commit is pushed to a repo (for example triggered by a post-commit hook).

I'll give you a little more context. We're interested in using Spor for two systems:

## Commenting system

We're building a system that will allow users to comment on sections of a file stored in Git, very much like Google Docs and Word lets you do it. The comments themselves will live in a database, and each comment will have a UID. This UID would also be assigned to the doc as spor metadata, allowing us to do the linking.

## Test Case Prioritisation

We're also considering building a system that would track the evolution of tests over time. This makes it possible to predict with high confidence what tests are most likely to fail after a modification to the source tree (using a trained naive bayes classifier).

We've previously looked into [lhdiff](https://asaduzzamanparvez.files.wordpress.com/2013/04/lhdiff-paper-long-version.pdf), but we're hoping spor can do the same for us.

---

Both these systems are collaborative in nature. The metadata needs to be shared, and will be updated by a central authority rather than by indivuduals.

**Comment by abingham** [comment]
Ok, I think I see what you're getting at. It sounds like you'll want spor
to the work of updating anchors/metadata as the tagged material changes,
but then you'll want to directly access the metadata database for the rest
of your operations. Is that about right? If so, I think that's a good basis
on which to start thinking about abstracting the idea of "storage backend"
for spor.

**Comment by aslakhellesoy** [comment]
Yep, that's right @abingham. I'm looping in @tooky who might have some additional thoughts.
